html/includes/common/top-devices.inc.php

```diff
@@ -85,46 +144,113 @@
+                      FROM `processors` AS `procs`, `devices` AS `d` 
+                      WHERE `d`.`device_id` = `procs`.`device_id` 
+                      AND unix_timestamp() - UNIX_TIMESTAMP(`last_polled`) < :interval 
+                      GROUP BY `d`.`device_id` 
+                      ORDER BY `cpuload` DESC 
+                      LIMIT :count';
+        } else {
+            $query = 'SELECT `hostname`, `last_polled`, `d`.`device_id`, avg(`processor_usage`) as `cpuload` 
+                      FROM `processors` AS procs, `devices` AS `d`, `devices_perms` AS `P`
+					  WHERE `P`.`user_id` = :user AND `P`.`device_id` = `procs`.`device_id` 
+                      AND unix_timestamp() - UNIX_TIMESTAMP(`last_polled`) < :interval
+                      GROUP BY `procs`.`device_id` 
+                      ORDER BY `cpuload` DESC
+                      LIMIT :count';
+        }
+    } elseif ($top_query == 4) {
+        $table_header = 'Memory usage';
+        if (is_admin() || is_read()) {
+            $query = 'SELECT `hostname`, `last_polled`, `d`.`device_id`, `mempool_perc` 
+                      FROM `mempools` as `mem`, `devices` as `d`
+                      WHERE `d`.`device_id` = `mem`.`device_id`
+                      AND `mempool_descr` IN (\'Physical memory\',\'Memory\')
+                      AND unix_timestamp() - UNIX_TIMESTAMP(`last_polled`) < :interval 
+                      ORDER BY `mempool_perc` DESC
+                      LIMIT :count';
+        } else {
+            $query = 'SELECT `hostname`, `last_polled`, `d`.`device_id`, `mempool_perc` 
+                      FROM `mempools` as `mem`, `devices` as `d`, `devices_perms` AS `P`
+                      WHERE `P`.`user_id` = :user AND `P`.`device_id` = `mem`.`device_id`
+                      AND `mempool_descr` IN (\'Physical memory\',\'Memory\')
+                      AND unix_timestamp() - UNIX_TIMESTAMP(`last_polled`) < :interval 
+                      ORDER BY `mempool_perc` DESC
+                      LIMIT :count';
+        }
     }
+
     $common_output[] = '
-<div class="table-responsive">
-<table class="table table-hover table-condensed table-striped bootgrid-table">
-  <thead>
-    <tr>
-      <th class="text-left">Device</th>
-      <th class="text-left">Total traffic</a></th>
-   </tr>
-  </thead>
-  <tbody>
-    ';
+    <div class="table-responsive">
+        <table class="table table-hover table-condensed table-striped bootgrid-table">
+        <thead>
+            <tr>
+                <th class="text-left">Device</th>
+                <th class="text-left">' . $table_header . '</a></th>
+            </tr>
+        </thead>
+        <tbody>';
 
     foreach (dbFetchRows($query, $params) as $result) {
+        if ($top_query == 0) {
+            $graphs = generate_device_link($result, generate_minigraph_image($result, $config['time']['day'], $config['time']['now'], 'device_bits', 'no', 150, 21), array(), 0, 0, 0);
```

Comment:
You could just set the relevant variables in these if checks and then run generate_minigraph_image at the end.